File: eco-core/core-plugin/src/main/java/com/willfp/eco/spigot/integrations/antigrief/AntigriefTowny.java
package com.willfp.eco.spigot.integrations.antigrief;

import com.palmergames.bukkit.towny.TownyAPI;
import com.palmergames.bukkit.towny.TownyUniverse;
import com.palmergames.bukkit.towny.object.Town;
import com.palmergames.bukkit.towny.object.TownyPermission;
import com.palmergames.bukkit.towny.object.TownyWorld;
import com.palmergames.bukkit.towny.object.WorldCoord;
import com.palmergames.bukkit.towny.utils.PlayerCacheUtil;
import com.willfp.eco.core.integrations.antigrief.AntigriefWrapper;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

public class AntigriefTowny implements AntigriefWrapper {
    @Override
    public boolean canBreakBlock(@NotNull final Player player,
                                 @NotNull final Block block) {
        TownyWorld world = TownyUniverse.getInstance().getWorldMap().get(block.getLocation().getWorld().getName());
        if (world == null) {
            return true;
        }
        if (TownyAPI.getInstance().isWilderness(block)) {
            return true;
        }
        return PlayerCacheUtil.getCachePermission(player, block.getLocation(), block.getType(), TownyPermission.ActionType.DESTROY)
                || PlayerCacheUtil.getCachePermission(player, block.getLocation(), block.getType(), TownyPermission.ActionType.BUILD);
    }

    @Override
    public boolean canCreateExplosion(@NotNull final Player player,
                                      @NotNull final Location location) {
        TownyWorld world = TownyUniverse.getInstance().getWorldMap().get(location.getWorld().getName());
        if (world == null) {
            return true;
        }
        if (TownyAPI.getInstance().isWilderness(location)) {
            return true;
        }
        return PlayerCacheUtil.getCachePermission(player, location, Material.TNT, TownyPermission.ActionType.ITEM_USE);
    }

    @Override
    public boolean canPlaceBlock(@NotNull final Player player,
                                 @NotNull final Block block) {
        TownyWorld world = TownyUniverse.getInstance().getWorldMap().get(block.getLocation().getWorld().getName());
        if (world == null) {
            return true;
        }
        if (TownyAPI.getInstance().isWilderness(block)) {
            return true;
        }
        return PlayerCacheUtil.getCachePermission(player, block.getLocation(), block.getType(), TownyPermission.ActionType.BUILD);
    }

    @Override
    public boolean canInjure(@NotNull final Player player,
                             @NotNull final LivingEntity victim) {
        TownyWorld world = TownyUniverse.getInstance().getWorldMap().get(victim.getLocation().getWorld().getName());
        if (world == null) {
            return true;
        }
        if (TownyAPI.getInstance().isWilderness(victim.getLocation())) {
            if (victim instanceof Player) {
                return world.isPVP();
            } else {
                return true;
            }
        }
        if (victim instanceof Player) {
            try {
                Town town = WorldCoord.parseWorldCoord(victim.getLocation()).getTownBlock().getTown();
                return town.isPVP();
            } catch (Exception ignored) {
                // If exception, no town was found, thus return true.
            }
        } else {
            try {
                Town town = WorldCoord.parseWorldCoord(victim.getLocation()).getTownBlock().getTown();
                return town.hasMobs();
            } catch (Exception ignored) {
                // If exception, no town was found, thus return true.
            }
        }
        return true;
    }

    @Override
    public String getPluginName() {
        return "Towny";
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/eco/spigot/integrations/antigrief/AntigriefTowny.kt
package com.willfp.eco.spigot.integrations.antigrief

import com.palmergames.bukkit.towny.TownyAPI
import com.palmergames.bukkit.towny.TownyUniverse
import com.palmergames.bukkit.towny.`object`.Town
import com.palmergames.bukkit.towny.`object`.TownyPermission
import com.palmergames.bukkit.towny.`object`.WorldCoord
import com.palmergames.bukkit.towny.utils.PlayerCacheUtil
import com.willfp.eco.core.integrations.antigrief.AntigriefWrapper
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player

class AntigriefTowny : AntigriefWrapper {
    override fun canBreakBlock(
        player: Player,
        block: Block
    ): Boolean {
        val world = TownyUniverse.getInstance().worldMap[block.location.world!!.name] ?: return true
        return if (TownyAPI.getInstance().isWilderness(block)) {
            true
        } else PlayerCacheUtil.getCachePermission(
            player,
            block.location,
            block.type,
            TownyPermission.ActionType.DESTROY
        )
                || PlayerCacheUtil.getCachePermission(
            player,
            block.location,
            block.type,
            TownyPermission.ActionType.BUILD
        )
    }

    override fun canCreateExplosion(
        player: Player,
        location: Location
    ): Boolean {
        val world = TownyUniverse.getInstance().worldMap[location.world!!.name] ?: return true
        return if (TownyAPI.getInstance().isWilderness(location)) {
            true
        } else PlayerCacheUtil.getCachePermission(player, location, Material.TNT, TownyPermission.ActionType.ITEM_USE)
    }

    override fun canPlaceBlock(
        player: Player,
        block: Block
    ): Boolean {
        val world = TownyUniverse.getInstance().worldMap[block.location.world!!.name] ?: return true
        return if (TownyAPI.getInstance().isWilderness(block)) {
            true
        } else PlayerCacheUtil.getCachePermission(player, block.location, block.type, TownyPermission.ActionType.BUILD)
    }

    override fun canInjure(
        player: Player,
        victim: LivingEntity
    ): Boolean {
        val world = TownyUniverse.getInstance().worldMap[victim.location.world!!.name] ?: return true
        if (TownyAPI.getInstance().isWilderness(victim.location)) {
            return if (victim is Player) {
                world.isPVP
            } else {
                true
            }
        }
        if (victim is Player) {
            try {
                val town: Town = WorldCoord.parseWorldCoord(victim.getLocation()).getTownBlock().getTown()
                return town.isPVP()
            } catch (ignored: Exception) {
                // If exception, no town was found, thus return true.
            }
        } else {
            try {
                val town: Town = WorldCoord.parseWorldCoord(victim.location).getTownBlock().getTown()
                return town.hasMobs()
            } catch (ignored: Exception) {
                // If exception, no town was found, thus return true.
            }
        }
        return true
    }

    override fun getPluginName(): String {
        return "Towny"
    }
}